### ISSUE TITLE
[API Platform] error / exception instead of a redirect

### ISSUE THREAD

**Issue by dkarlovi** [rollerworks/search#174]
| Q                     | A
| --------------------- | -----
| Bug report?           | no
| Feature request?      | yes
| BC Break report?      | no
| RFC?                  | yes

_(first of all, sorry about the issue storm)_

Currently, if a search query is not valid or was reduced, you do a redirect to a proper query. This makes discovering what I did wrong as the API user much harder (as it might redirect you to no search query at all).

What I suggest is

1. throw an exception which describes why the query was wrong instead if it can't be normalized or doesn't match expectations (an error)
2. provide a [`canonical` header](https://en.wikipedia.org/wiki/Canonical_link_element) if the query was simplified, but still able to apply (warning)

This would allow users to discover API features just by using it, to also disover optimal queries, but still use it without implementing "follow redirects" etc.

**Comment by sstok** [comment]
This has my vote. But needs a bit more work as you can't set a `canonical` header in `SearchConditionListener`. This will stop propagation to other listeners (basically breaking the event flow).

Instead we need the to use the Request attributes to set a `canonical` location, and then in another listener set this for the response. Or is there a better solution?

**Comment by sstok** [comment]
Did you actually encounter situations were an invalid query resulted in an empty redirect? 

The main reason for a redirect is when the search-condition was optimized you don't want to keep using the original version as this makes caching a nightmare and could potentially lead to a Dos.

**Comment by dkarlovi** [comment]
> Did you actually encounter situations were an invalid query resulted in an empty redirect?

Yes, as I've described in #175. 

> this makes caching a nightmare and could potentially lead to a Dos.

How so? You cache the unoptimized version too, it should evaluate to the same outcome as the optimized version.

**Comment by sstok** [comment]
Good news, I found a way to solve this. But it might shock you ⚠️ .


Remove the Optimizers! I don't know what I was thinking when I introduced this 😅 the chance of finding an optimization is really small (because the optimizers don't work with deeper levels) and the overhead caused by this is not worth the gain.

I also have a number of other ideas to improve the overall system, but I need to work out some of the details 😛 

Btw, are you caching the input-processed SearchCondition or are you processing the input for every request?
